runtime/src/accounts_db.rs

```diff
@@ -1934,6 +1939,46 @@ impl AccountsDb {
         });
         purge_filter.stop();
 
+        // When using incremental snapshots, do not purge zero-lamport accounts if the slot is
+        // higher than the last full snapshot slot.  This is to protect against the following
+        // scenario:
+        //
+        //      A full snapshot is taken, and it contains an account with a non-zero balance.
+        //      Later, that account's  goes to zero.  Evntually cleaning runs, and before, this
+        //      account would be cleaned up.  Finally, an incremental snapshot is taken.
+        //
+        //      Later, the incremental (and full) snapshot is used to rebuild the bank and accounts
+        //      database (e.x. if the node restarts).  The full snapshot _does_ contain the account
+        //      (from above) and its balance is non-zero, however, since the account was cleaned up
+        //      in a later slot, the incremental snapshot does not contain any info about this
+        //      account, thus, the accounts database will contain the old info from this account,
+        //      which has its old non-zero balance.  Very bad!
+        //
+        // This filtering step can be skipped if there is no `last_full_snapshot_slot`, or if the
+        // `max_clean_root` is less-than-or-equal-to the `last_full_snapshot_slot`.
+        let mut measure_purge_filter_zero_lamports =
+            Measure::start("filter purges w.r.t. snapshots");
+        if let Some(last_full_snapshot_slot) = last_full_snapshot_slot {
+            if max_clean_root.unwrap_or(Slot::MAX) > last_full_snapshot_slot {
+                purges_zero_lamports.retain(|_pubkey, (slot_account_infos, _ref_count)| {
+                    let slot_account_info_at_highest_slot = slot_account_infos
+                        .iter()
+                        .max_by_key(|(slot, _account_info)| slot);
+
+                    slot_account_info_at_highest_slot.map_or(true, |(slot, account_info)| {
+                        // Do *not* purge zero-lamport account if the slot is greater than the last
+                        // full snapshot slot.  Since we're `retain`ing the accounts-to-purge, I
+                        // felt creating the `cannot_purge` variable made this easier to
+                        // understand.
+                        let cannot_purge =
+                            account_info.is_zero_lamport() && *slot > last_full_snapshot_slot;
+                        !cannot_purge
+                    })
+                });
+            }
+        }
+        measure_purge_filter_zero_lamports.stop();
+
```

Comment:
let's extract this into a separate function behind a feature flag for incremental snapshots so that we can toggle off incremental snapshot cleaning in case we need to roll this back.